File: src/panda/scala/linkedlist/RemoveLinkedListElement.scala
package panda.scala.linkedlist

object RemoveLinkedListElement {
  def removeElements(head: ListNode, `val`: Int): ListNode = {
    var ahead = head
    while (ahead != null && ahead._x == `val`) {
      ahead = ahead.next
    }
    if (ahead == null) {
      return head
    }

    var prev = ahead

    while (prev.next != null) {
      if (prev.next._x == `val`) {
        prev.next = prev.next.next
      } else {
        prev = prev.next
      }
    }
    ahead
  }

  def removeElements2(head: ListNode, `val`: Int): ListNode = {
    if (head == null) {
      return null
    }
    head.next = removeElements2(head.next, `val`)
    //    if(head._x == `val`){
    //      return res
    //    }else{
    //      head.next = res
    //      return head
    //    }

    if (head._x == `val`) head.next else head
  }


}

File: src/panda/java/linkedlist/RemoveLinkedListElement.java
package panda.java.linkedlist;

/**
 * 删除链表中等于给定值 val 的所有节点。
 */

public class RemoveLinkedListElement {
    public static ListNode removeElements(ListNode head, int val) {
        //删除开始部分的元素
        while (head != null && head.val == val) {
            head = head.next;
        }
        //删除完了
        if (head == null) {
            return null;
        }
        ListNode prev = head;
        while (prev.next != null) {
            if (prev.next.val == val) {
                prev.next = prev.next.next;
            } else {
                prev = prev.next;
            }
        }
        return head;
    }
}
